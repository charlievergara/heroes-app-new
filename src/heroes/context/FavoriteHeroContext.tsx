import { createContext, useEffect, useState, type PropsWithChildren } from "react";
import type { Hero } from "../types/hero.interface";


interface FavoriteContext {
    favorites: Hero[];
    favoriteCount:number;
    
    //methods
    isFavorite: (hero:Hero) => boolean;
    toggleFavorite: (hero:Hero) => void;

}

export const FavoriteHeroContext = createContext({} as FavoriteContext)

const getFavoritesFromLocalStorage = ():Hero[]=> {
    const favorites = localStorage.getItem('favorites')

    return favorites ? JSON.parse(favorites): []
}

export const FavoriteHeroProvider = ({children}:PropsWithChildren) => {

    const [favorites, setFavorites] = useState<Hero[]>(getFavoritesFromLocalStorage())

    const toggleFavorite = (hero: Hero) => {
        
        // const heroExist = favorites.find((currentHero:Hero) => currentHero.id === hero.id)

        if(isFavorite(hero)){
            const newFavorites = favorites.filter(h=>h.id !== hero.id)
            setFavorites(newFavorites)
        }else{
            setFavorites([...favorites, hero])
        }
    }

    const isFavorite = (hero:Hero) => {
        return favorites.some((currentHero:Hero) => currentHero.id === hero.id)
    }

    useEffect(() => {
      localStorage.setItem('favorites', JSON.stringify(favorites))
    }, [favorites])
    

  return (
    <FavoriteHeroContext value={{
        favoriteCount:favorites.length,
        favorites,
        toggleFavorite,
        isFavorite
    }}>
        {children}
    </FavoriteHeroContext>
  )
}
