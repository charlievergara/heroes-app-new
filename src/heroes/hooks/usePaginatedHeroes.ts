import { useQuery } from "@tanstack/react-query"
import { getHeroesByPageAction } from "../actions/get-heores-by-page.action"

export const usePaginatedHeroes = (page:string, limit:string, category:string = 'all') => {
    return useQuery({
        queryKey:['heroes', {page, limit,category}],
        queryFn:()=> getHeroesByPageAction(+page, +limit, category),
        staleTime:1000*60*5,
      })
}
