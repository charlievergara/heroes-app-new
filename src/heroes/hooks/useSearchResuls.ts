import { useQuery } from '@tanstack/react-query'
import { getResultsSearchAction, type Options } from '../actions/get-results-search.action'

export const useSearchResuls = (options:Options) => {
 return useQuery({
    queryKey:['heroes','search', {...options}],
    queryFn:()=> getResultsSearchAction(options),
    staleTime: 1000*60*5
 }) 
}
