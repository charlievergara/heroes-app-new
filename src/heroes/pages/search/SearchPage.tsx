import { Jumbotron } from "@/components/custom/Jumbotron"
import { HeroStats } from "@/heroes/components/HeroStats"
import { SearchControls } from "./ui/SearchControls"
import { MainBreadCrumb } from "@/components/custom/MainBreadCrumb"
import { HeroGrid } from "@/heroes/components/HeroGrid"
import { useSearchParams } from "react-router"
import { useSearchResuls } from "@/heroes/hooks/useSearchResuls"

export const SearchPage = () => {

  const breadcrumbs = [
    {label:'Home', to:'/'},
    {label:'Search Heroes', to:'/search'},
  ]

  const [searchParams, _] = useSearchParams()

  const search = searchParams.get('search') || ''
  const strength = searchParams.get('strength') ?? undefined

  const { data:heroesList } = useSearchResuls( {name:search, strength} );

  return (
    <>
    <Jumbotron title="Buscar superheroes" description="Descubre, explora y administra Superheroes y villanos"></Jumbotron>

    <MainBreadCrumb currentPage="Search heores" breadcrumbs={breadcrumbs}></MainBreadCrumb>

    {/* Stats Dashboard */}
    <HeroStats></HeroStats>

    {/* Filter and search controls */}
    <SearchControls></SearchControls>

    <HeroGrid heroList={heroesList?.heroes ?? []} ></HeroGrid>
    </>

  )
}

export default SearchPage