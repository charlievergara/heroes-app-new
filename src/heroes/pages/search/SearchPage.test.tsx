import { render, screen, waitFor } from "@testing-library/react";
import { beforeEach, describe, expect, test, vi } from "vitest";
import SearchPage from "./SearchPage";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { FavoriteHeroProvider } from "@/heroes/context/FavoriteHeroContext";
import { MemoryRouter } from "react-router";
import { useSearchResuls } from "@/heroes/hooks/useSearchResuls";
import type { Hero } from '../../types/hero.interface';

vi.mock('@/heroes/hooks/useSearchResuls')

const mockedUseSearchResuls = vi.mocked(useSearchResuls)

mockedUseSearchResuls.mockReturnValue({
    data:{heroes:[{id:'1', name:'Carlos Kent'}]},
    isLoading: false,
    isError:false,
    isSuccess: true
} as unknown as ReturnType<typeof useSearchResuls>)

vi.mock('@/components/custom/Jumbotron', ()=>({
    Jumbotron:()=><div data-testid="custom-jumbotron"></div>
}))

vi.mock('@/heroes/components/HeroGrid', ()=>({
    HeroGrid:({heroList}:{heroList:Hero[]})=>(<div data-testid="hero-grid">
        {heroList.map((hero)=>(<div key={hero.id}>{hero.name}</div>))}
    </div>)
}))


const queryClient = new QueryClient()

const renderPage = (initialEntries:string[] = ['/search'])=>{
    return render(
        <MemoryRouter initialEntries={initialEntries}>
            <FavoriteHeroProvider>
                <QueryClientProvider client={queryClient}>
                        <SearchPage></SearchPage>
                </QueryClientProvider>
            </FavoriteHeroProvider>
        </MemoryRouter>
    )
}


describe('SearchPage', () => { 

    beforeEach(()=>{
        vi.clearAllMocks();
        queryClient.clear();
    })

    test('should render search page with default values', () => { 
        const {container} = renderPage()
        expect(mockedUseSearchResuls).toHaveBeenCalledWith({name:'', strength:undefined})

        expect(container).toMatchSnapshot()
     })

     test('should query hook with name parameter', () => { 
        const {container} = renderPage(['/search?search=superman'])
        expect(mockedUseSearchResuls).toHaveBeenCalledWith({name:'superman', strength:undefined})
        
    })
     test('should query hook with strength parameter', () => { 
        const {container} = renderPage(['/search?strength=6'])
        expect(mockedUseSearchResuls).toHaveBeenCalledWith({name:'', strength:"6"})  
    })

    test('should render heroGrid with search Results', () => { 
        const {container} = renderPage()
            
        const heroGrid = screen.getByTestId('hero-grid')
        expect(screen.getByText('Carlos Kent')).toBeDefined()
     })
 })