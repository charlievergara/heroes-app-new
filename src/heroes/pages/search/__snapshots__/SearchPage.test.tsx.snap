// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchPage > should render search page with default values 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <div
        class="flex items-center"
      >
        <li
          aria-hidden="true"
          class="[&>svg]:size-3.5"
          data-slot="breadcrumb-separator"
          role="presentation"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-right"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m9 18 6-6-6-6"
            />
          </svg>
        </li>
        <li
          class="inline-flex items-center gap-1.5"
          data-slot="breadcrumb-item"
        >
          <a
            class="hover:text-foreground transition-colors"
            data-discover="true"
            data-slot="breadcrumb-link"
            href="/"
          >
            Home
          </a>
        </li>
      </div>
      <div
        class="flex items-center"
      >
        <li
          aria-hidden="true"
          class="[&>svg]:size-3.5"
          data-slot="breadcrumb-separator"
          role="presentation"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-right"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m9 18 6-6-6-6"
            />
          </svg>
        </li>
        <li
          class="inline-flex items-center gap-1.5"
          data-slot="breadcrumb-item"
        >
          <a
            class="hover:text-foreground transition-colors"
            data-discover="true"
            data-slot="breadcrumb-link"
            href="/search"
          >
            Search Heroes
          </a>
        </li>
      </div>
    </ol>
  </nav>
  <div
    class="grid grid-cols-2 md:grid-cols-4 gap-4 mb-8"
  >
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Total Characters
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-users h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 21v-2a4 4 0 0 0-4-4H6a4 4 0 0 0-4 4v2"
          />
          <path
            d="M16 3.128a4 4 0 0 1 0 7.744"
          />
          <path
            d="M22 21v-2a4 4 0 0 0-3-3.87"
          />
          <circle
            cx="9"
            cy="7"
            r="4"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-2xl font-bold"
        />
        <div
          class="flex gap-1 mt-2"
        >
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 text-xs"
            data-slot="badge"
          >
             Heroes
          </span>
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-destructive text-white [a&]:hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60 text-xs"
            data-slot="badge"
          >
             Villains
          </span>
        </div>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Favoritos
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-heart h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 9.5a5.5 5.5 0 0 1 9.591-3.676.56.56 0 0 0 .818 0A5.49 5.49 0 0 1 22 9.5c0 2.29-1.5 4-3 5.5l-5.492 5.313a2 2 0 0 1-3 .019L5 15c-1.5-1.5-3-3.2-3-5.5"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-2xl font-bold text-red-600"
          data-testid="favoriteCount"
        >
          0
        </div>
        <p
          class="text-xs text-muted-foreground"
          data-testid="favorite-content"
        >
          Loading...
        </p>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Strongest
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-zap h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 14a1 1 0 0 1-.78-1.63l9.9-10.2a.5.5 0 0 1 .86.46l-1.92 6.02A1 1 0 0 0 13 10h7a1 1 0 0 1 .78 1.63l-9.9 10.2a.5.5 0 0 1-.86-.46l1.92-6.02A1 1 0 0 0 11 14z"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-lg font-bold"
        />
        <p
          class="text-xs text-muted-foreground"
        >
          Strength: 
          /10
        </p>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Smartest
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-trophy h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 14.66v1.626a2 2 0 0 1-.976 1.696A5 5 0 0 0 7 21.978"
          />
          <path
            d="M14 14.66v1.626a2 2 0 0 0 .976 1.696A5 5 0 0 1 17 21.978"
          />
          <path
            d="M18 9h1.5a1 1 0 0 0 0-5H18"
          />
          <path
            d="M4 22h16"
          />
          <path
            d="M6 9a6 6 0 0 0 12 0V3a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1z"
          />
          <path
            d="M6 9H4.5a1 1 0 0 1 0-5H6"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-lg font-bold"
        />
        <p
          class="text-xs text-muted-foreground"
        >
          Intelligence: 
          /10
        </p>
      </div>
    </div>
  </div>
  <div
    class="flex flex-col lg:flex-row gap-4 mb-8"
  >
    <div
      class="relative flex-1"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-search absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21 21-4.34-4.34"
        />
        <circle
          cx="11"
          cy="11"
          r="8"
        />
      </svg>
      <input
        class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex w-full min-w-0 rounded-md border px-3 py-1 shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive pl-12 h-12 text-lg bg-white"
        data-slot="input"
        data-testid="searchInput"
        placeholder="Search heroes, villains, powers, teams..."
        value=""
      />
    </div>
    <div
      class="flex gap-2"
    >
      <button
        class="cursor-pointer inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 px-4 py-2 has-[>svg]:px-3 h-12"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-funnel h-4 w-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 20a1 1 0 0 0 .553.895l2 1A1 1 0 0 0 14 21v-7a2 2 0 0 1 .517-1.341L21.74 4.67A1 1 0 0 0 21 3H3a1 1 0 0 0-.742 1.67l7.225 7.989A2 2 0 0 1 10 14z"
          />
        </svg>
        Filters
      </button>
      <button
        class="cursor-pointer inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 px-4 py-2 has-[>svg]:px-3 h-12 bg-transparent"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-arrow-up-narrow-wide h-4 w-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3 8 4-4 4 4"
          />
          <path
            d="M7 4v16"
          />
          <path
            d="M11 12h4"
          />
          <path
            d="M11 16h7"
          />
          <path
            d="M11 20h10"
          />
        </svg>
        Sort by Name
      </button>
      <button
        class="cursor-pointer inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 px-4 py-2 has-[>svg]:px-3 h-12 bg-transparent"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-grid3x3 lucide-grid-3x3 h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="18"
            rx="2"
            width="18"
            x="3"
            y="3"
          />
          <path
            d="M3 9h18"
          />
          <path
            d="M3 15h18"
          />
          <path
            d="M9 3v18"
          />
          <path
            d="M15 3v18"
          />
        </svg>
      </button>
      <button
        class="cursor-pointer inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 px-4 py-2 has-[>svg]:px-3 h-12"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-plus h-4 w-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="M12 5v14"
          />
        </svg>
        Add Character
      </button>
    </div>
  </div>
  <div
    data-orientation="vertical"
    data-slot="accordion"
    data-testid="accordion"
  >
    <div
      class="border-b last:border-b-0"
      data-orientation="vertical"
      data-slot="accordion-item"
      data-state="closed"
    >
      <div
        aria-labelledby="radix-«r0»"
        class="data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down overflow-hidden text-sm"
        data-orientation="vertical"
        data-slot="accordion-content"
        data-state="closed"
        hidden=""
        id="radix-«r1»"
        role="region"
        style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
      />
    </div>
  </div>
  <div
    data-testid="hero-grid"
  >
    <div>
      Carlos Kent
    </div>
  </div>
</div>
`;
